Mathlib/Combinatorics/SimpleGraph/Diameter.lean

```diff
@@ -0,0 +1,218 @@
+
+lemma ediam_mono_of_finite [Nonempty α] [Finite α] (h : G ≤ G') :
+    G'.ediam ≤ G.ediam :=
+  have ⟨_, _, huv⟩ := G'.exists_edist_eq_ediam_of_finite
+  huv ▸ (edist_le_subgraph_edist h).trans edist_le_ediam
+
+@[simp]
+lemma ediam_bot [Nontrivial α] : (⊥ : SimpleGraph α).ediam = ⊤ :=
+  ediam_eq_top_of_not_connected bot_not_connected
+
+@[simp]
+lemma ediam_top [Nontrivial α] : (⊤ : SimpleGraph α).ediam = 1 := by
+  apply le_antisymm ?_ <| ENat.one_le_iff_pos.mpr <| pos_iff_ne_zero.mpr ediam_ne_zero
+  apply ediam_def ▸ iSup_le_iff.mpr
+  intro p
+  by_cases h : (⊤ : SimpleGraph α).Adj p.1 p.2
+  · apply le_of_eq <| edist_eq_one_iff_adj.mpr h
+  · simp_all
+
+@[simp]
+lemma ediam_eq_one [Nontrivial α] : G.ediam = 1 ↔ G = ⊤ := by
+  refine ⟨fun h₁ ↦ ?_, fun h ↦ h ▸ ediam_top⟩
+  ext u v
+  refine ⟨fun h ↦ h.ne, fun h₂ ↦ ?_⟩
+  apply G.edist_pos_of_ne at h₂
+  apply le_of_eq at h₁
+  rw [ediam_def, iSup_le_iff] at h₁
+  exact edist_eq_one_iff_adj.mp <| le_antisymm (h₁ (u, v)) <| ENat.one_le_iff_pos.mpr h₂
+
+end ediam
+
+section diam
+
+/--
+The diameter is the greatest distance between any two vertices, with the value `0` in
+case the distances are not bounded above, or the graph is not connected.
+-/
+noncomputable def diam (G : SimpleGraph α) :=
+  G.ediam.toNat
+
+lemma diam_def : G.diam = (⨆ p : α × α, G.edist p.1 p.2).toNat := by
+  rw [diam, ediam_def]
+
+lemma dist_le_diam (h : G.ediam ≠ ⊤) {u v : α} : G.dist u v ≤ G.diam :=
+  ENat.toNat_le_toNat edist_le_ediam h
+
+lemma nontrivial_of_diam_ne_zero (h : G.diam ≠ 0) : Nontrivial α := by
+  apply G.nontrivial_of_ediam_ne_zero
+  contrapose! h
+  simp [diam, h]
+
+lemma diam_eq_zero_of_not_connected (h : ¬G.Connected) : G.diam = 0 := by
+  cases isEmpty_or_nonempty α
+  · rw [diam, ediam, ciSup_of_empty, bot_eq_zero']; rfl
+  · rw [diam, ediam_eq_top_of_not_connected h, ENat.toNat_top]
+
+lemma ediam_ne_top_of_diam_ne_zero (h : G.diam ≠ 0) : G.ediam ≠ ⊤ :=
+  Ne.symm <| ne_of_apply_ne ENat.toNat fun a ↦ h <| id a.symm
+
+lemma diam_eq_zero_of_ediam_ne_top (h : G.ediam = ⊤) : G.diam = 0 :=
+  ediam_ne_top_of_diam_ne_zero.mtr h
+
+lemma exists_dist_eq_diam_of_ne_zero [Nonempty α] :
+    ∃ u v, G.dist u v = G.diam := by
+  by_cases h : G.diam = 0
+  · simp [h]
+  · obtain ⟨u, v, huv⟩ := exists_edist_eq_ediam_of_ne_top <| ediam_ne_top_of_diam_ne_zero h
+    use u, v
+    rw [diam, dist, congrArg ENat.toNat huv]
+
+lemma exists_dist_eq_diam_of_finite [Nonempty α] [Finite α] :
+    ∃ u v, G.dist u v = G.diam := by
+  obtain ⟨u, v, huv⟩ := G.exists_edist_eq_ediam_of_finite
+  use u, v
+  rw [diam, dist, congrArg ENat.toNat huv]
+
+lemma diam_mono_of_ne_zero (h : G ≤ G') (hn : G.diam ≠ 0) :
+    G'.diam ≤ G.diam := by
+  by_cases hn' : G'.diam = 0
+  · omega
+  · have : Nontrivial α := nontrivial_of_diam_ne_zero hn
+    exact ENat.toNat_le_toNat (ediam_mono_of_ne_top h (ediam_ne_top_of_diam_ne_zero hn'))
+      <| ediam_ne_top_of_diam_ne_zero hn
```

Comment:
Similarly here you worked too hard, probably because you overstrengthened `hn`
```suggestion
lemma diam_mono_of_ediam_ne_top (h : G ≤ G') (hn : G.ediam ≠ ⊤) : G'.diam ≤ G.diam :=
  ENat.toNat_le_toNat (ediam_mono h) hn
```